quickwit/quickwit-indexing/src/source/doc_file_reader.rs

```diff
@@ -129,6 +135,142 @@ impl DocFileReader {
+        uri: &Uri,
+        position: Position,
+    ) -> anyhow::Result<Self> {
+        let current_offset = match position {
+            Position::Beginning => 0,
+            Position::Offset(offset) => offset
+                .as_usize()
+                .context("file offset should be stored as usize")?,
+            Position::Eof(_) => {
+                return Ok(ObjectUriBatchReader {
+                    partition_id,
+                    reader: DocFileReader::empty(),
+                    current_offset: 0,
+                    is_eof: true,
+                })
+            }
+        };
+        let reader = DocFileReader::from_uri(storage_resolver, uri, current_offset).await?;
+        Ok(ObjectUriBatchReader {
+            partition_id,
+            reader,
+            current_offset,
+            is_eof: false,
+        })
+    }
+}
+
+#[async_trait]
+impl BatchReader for ObjectUriBatchReader {
+    async fn read_batch(
+        &mut self,
+        source_ctx: &SourceContext,
+        source_type: SourceType,
+    ) -> anyhow::Result<BatchBuilder> {
+        let limit_num_bytes = self.current_offset + BATCH_NUM_BYTES_LIMIT as usize;
+        let mut new_offset = self.current_offset;
+        let mut batch_builder = BatchBuilder::new(source_type);
+        while new_offset < limit_num_bytes {
+            if let Some(record) = source_ctx.protect_future(self.reader.next_record()).await? {
+                new_offset = record.next_offset as usize;
+                batch_builder.add_doc(record.doc);
+            } else {
+                self.is_eof = true;
+                break;
+            }
+        }
+        if new_offset > self.current_offset {
+            let to_position = if self.is_eof {
+                Position::eof(new_offset)
+            } else {
+                Position::offset(new_offset)
+            };
+            batch_builder
+                .checkpoint_delta
+                .record_partition_delta(
+                    self.partition_id.clone(),
+                    Position::offset(self.current_offset),
+                    to_position,
+                )
+                .unwrap();
+            self.current_offset = new_offset;
+        }
+
+        Ok(batch_builder)
+    }
+
+    fn is_eof(&self) -> bool {
+        self.is_eof
+    }
+}
+
+pub struct StdinBatchReader {
+    reader: BufReader<tokio::io::Stdin>,
+    is_eof: bool,
+}
+
+impl StdinBatchReader {
+    pub fn new() -> Self {
+        Self {
+            reader: BufReader::new(tokio::io::stdin()),
+            is_eof: false,
+        }
+    }
+}
+
+#[async_trait]
+impl BatchReader for StdinBatchReader {
+    async fn read_batch(
+        &mut self,
+        source_ctx: &SourceContext,
+        source_type: SourceType,
+    ) -> anyhow::Result<BatchBuilder> {
+        let mut batch_builder = BatchBuilder::new(source_type);
+        while batch_builder.num_bytes < BATCH_NUM_BYTES_LIMIT {
+            let mut buf = String::new();
```

Comment:
why allocate at every single loop iteration?

Reply:
Copied from main https://github.com/quickwit-oss/quickwit/blob/84572ea5955ecbe2b0d57ae5a1fbe5d5600ce716/quickwit/quickwit-indexing/src/source/file_source.rs#L80. I guess we could allocate one big memory region that is `BATCH_NUM_BYTES_LIMIT`, load data into it and slice it by row. I think it's a bit far from this PR's concern though. Do you think these allocations are costly enough for opening an issue?